k2/csrc/determinize.cc

```diff
@@ -125,18 +258,169 @@ struct DetState {
+    bool inserted = p.second;
+    if (!inserted) {
+      DetStateElement *old_elem = p.first->second;
+      if (old_elem->weight > elem->weight) {  // old weight is better
+        this->RemoveElement(elem);
+      } else {
+        p.first->second = elem;
+        this->RemoveElement(old_elem);
+      }
+    }
+  }
+}
+
+void DetState::RemoveCommonPrefix(const Fsa &input_fsa,
+                                  const float *input_fsa_weights,
+                                  float *removed_weight_out,
+                                  std::vector<int32_t> *input_arcs) {
+
+  CHECK_GE(seq_len, 0);
+  int32_t len;
+  auto first_path = elements.front().path,
+      last_path = elements.back().path;
+
+  for (len = 1; len < seq_len; ++len) {
+    first_path = first_path->prev;
+    last_path = last_path->prev;
+    if (first_path == last_path) {
+      // Note: we are comparing pointers here.  We reached the same PathElement,
+      // which means we reached the same state.
+      break;
+    }
+  }
+  input_arcs->clear();
+  if (len < seq_len) {
+    /* We reach a common state after traversing fewer than `seq_len` arcs,
+       so we can remove a shared prefix. */
+    double removed_weight = 0.0;
+    int32_t new_seq_len = len,
+        removed_seq_len = seq_len - len;
+    input_arcs->resize(removed_seq_len);
+    // Advance base_state
+    int32_t new_base_state = input_fsa.arcs[first_path->arc_index].src_state;
+    for (; len < seq_len; ++len) {
+      auto arc = input_fsa.arcs[first_path->arc_index];
+      input_arcs[seq_len - 1 - len] = first_path->arc_index;
+      removed_weight += input_fsa_weights[first_path->arc_index];
+      first_path = first_path->prev;
+    }
+    // Check that we got to base_state.
+    CHECK((self->base_state == 0 && first_path == nullptr) ||
+          fsa.arcs[first_path->arc_index].dest_state == this->base_state);
+    this->base_state = new_base_state;
+    if (removed_weight != 0) {
+      for (DetStateElement &det_state_elem: elements) {
+        det_state_elem.weight -= removed_weight;
+      }
+    }
+    *removed_weight_out = removed_weight;
+  } else {
+    *removed_weight_out = 0;
+    input_arcs->clear();
+  }
+}
+
+void DetState::CheckElementOrder(const Fsa &input_fsa) const {
+  // Checks that the DetStateElements are in a lexicographical order on the
+  // lists of states in their paths.  This will be true becase of how we
+  // construct them (it requires on the IsArcSorted() property, whereby arcs
+  // leaving each state in the FSA are sorted first on label and then on
+  // dest_state.
+  if (seq_len == 0) {
+    CHECK(elements.size() == 1);
+    CHECK(elements.front().weight == 0.0);
+  }
+
+  std::vector<int32> prev_seq;
+  for (auto iter = elements.begin(); iter != elements.end(); ++iter) {
+    auto path = iter->path;
+    std::vector<int32> cur_seq;
+    for (int32_t i = 0; i < seq_len; i++) {
+      cur_seq.push_back(input_fsa.arcs[path->arc_index].prev_state);
```

Comment:
I got stuck. why `Fsa.arcs` has a property of `preve_state` as it's not `LogSumTracebackLink`

Reply:
should be src_state.   This is all an early draft.